File: client/Assets/UIDataBinding/databinding/CCActiveBInd.ts

import * as cc from "cc";
import { _decorator } from 'cc';
import { CCDataBindBase } from "./CCDataBindBase";
const { ccclass, property, menu } = _decorator;

@ccclass('CCActiveBind')
@menu("DataDrive/CCActiveBind")
export class CCActiveBind extends CCDataBindBase {
	@property({
		displayName: "可见性",
	})
	visible: string = ""

	protected needAttach() {
		if (this.visible) {
			return this.isAttachCalled && this.enabled && !!this.node.parent?.activeInHierarchy
		} else {
			return this.isAttachCalled && this.enabledInHierarchy
		}
	}

	/**
	 * 更新显示状态
	 */
	protected onBindItems() {
		this.checkVisible()
	}

	checkVisible() {
		if (!this.visible) {
			return false
		}
		let node = this.node
		if (!node) {
			return false
		}
		this.watchValueChange<boolean>(this.visible, (newValue) => {
			if (cc.isValid(node, true)) node.active = newValue;
		})
		return true
	}

}

File: client/Assets/UIDataBinding/databinding/CCActiveBInd.cs
using UnityEngine;
using System;
using System.Collections.Generic;

namespace DataBinding.UIBind
{
	[AddComponentMenu("DataDrive/CCActiveBind")]

	public class CCActiveBind : CCDataBindBase
	{
		[Rename("可见性")]
		public string visible = "";

		protected override bool needAttach()
		{
			if (!string.IsNullOrEmpty(this.visible))
			{
				if(this.isAttachCalled && this.enabled)
                {
					var p = this.transform.parent;
					return p != null ? p.IsActiveInHierarchy() : false;
				}
				return false;
			}
			else
			{
				return this.isAttachCalled && this.enabledInHierarchy;
			}
		}

		/**
		 * 更新显示状态
		 */
		protected override void onBindItems()
		{
			this.checkVisible();
		}

		public virtual bool checkVisible()
		{
			if (string.IsNullOrEmpty(this.visible))
			{
				return false;
			}
			var node = this.transform;
			if (node == null)
			{
				return false;
			}
			this.watchValueChange<bool>(this.visible, (object newValue, object oldValue) =>
			{
				if (Utils.isValid(node, true))
				{
					node.gameObject.SetActive(Utils.IsTrue(newValue));
				}
			});
			return true;
		}

	}
}
